pkg/controller/seed-controller-manager/cni-application-installation-controller/controller.go

```diff
@@ -0,0 +1,342 @@
+	"go.uber.org/zap"
+
+	appskubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/apps.kubermatic/v1"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	kubermaticv1helper "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1/helper"
+	clusterclient "k8c.io/kubermatic/v2/pkg/cluster/client"
+	"k8c.io/kubermatic/v2/pkg/cni"
+	kubermaticpred "k8c.io/kubermatic/v2/pkg/controller/util/predicate"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8c.io/kubermatic/v2/pkg/util/workerlabel"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "kkp-cni-application-installation-controller"
+
+	cniPluginNamespace = "kube-system"
+)
+
+// UserClusterClientProvider provides functionality to get a user cluster client.
+type UserClusterClientProvider interface {
+	GetClient(ctx context.Context, c *kubermaticv1.Cluster, options ...clusterclient.ConfigOption) (ctrlruntimeclient.Client, error)
+}
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName                    string
+	recorder                      record.EventRecorder
+	userClusterConnectionProvider UserClusterClientProvider
+	log                           *zap.SugaredLogger
+	versions                      kubermatic.Versions
+	overwriteRegistry             string
+}
+
+func Add(ctx context.Context, mgr manager.Manager, numWorkers int, workerName string, userClusterConnectionProvider UserClusterClientProvider, log *zap.SugaredLogger, versions kubermatic.Versions, overwriteRegistry string) error {
+	reconciler := &Reconciler{
+		Client:                        mgr.GetClient(),
+		workerName:                    workerName,
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log.Named(ControllerName),
+		versions:                      versions,
+		overwriteRegistry:             overwriteRegistry,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	// Predicate for filtering out events for clusters whose CNI is not managed by Applications infra
+	cniManagedByAppInfraPredicate := kubermaticpred.Factory(func(o ctrlruntimeclient.Object) bool {
```

Comment:
For UPDATE events, `kubermaticpred.Factory`  apply the filter to both the old and new object and OR's the result.

Meaning that if the user does a manual CNI migration from a CNI managed by app to a CNI managed by addons it will:
* uninstall the "new" CNI managed by addons
* try to install the "new" CNI  by creating an application CR that references an ApplicationDefinition that does not exist 

So for UPDATE events, this predicate should only check that the new object is managed by app.

Reply:
good point, thanks for the catch!